vendor/k8s.io/kube-openapi/pkg/generators/fieldversion.go

```diff
@@ -0,0 +1,125 @@
+// version for the prerelease and feature gate.
+// TODO: update docs
+const fieldVersionExtension = "x-kubernetes-field-version"
+
+// Example tag format:
+// +k8s:openapi-gen:prerelease=alpha,version=v1.20,featureGate=EndpointSliceNodeName
+const tagPrerelease = "k8s:openapi-gen:prerelease"
+
+var allowedPrereleaseNames = sets.NewString("alpha", "beta", "stable")
+
+// TODO: add docs
+type fieldVersion struct {
+	prerelease  string
+	version     string
+	featureGate string
+}
+
+func parseFieldVersions(comments []string) ([]fieldVersion, []error) {
+	fieldVersions := []fieldVersion{}
+	errors := []error{}
+
+	for _, comment := range types.ExtractCommentTags("+", comments)[tagPrerelease] {
+		itemMap := map[string]string{}
+		itemValues := strings.Split(comment, ",")
+
+		prereleaseName := strings.TrimSpace(itemValues[0])
+		if !allowedPrereleaseNames.Has(prereleaseName) {
+			errors = append(errors, fmt.Errorf("unrecognized prerelease: %q. supported names are: %q", prereleaseName, allowedPrereleaseNames.List()))
+		}
+
+		for _, itemValue := range itemValues[1:] {
+			itemValueParts := strings.Split(itemValue, "=")
+			if len(itemValueParts) == 2 {
+				itemMap[strings.TrimSpace(itemValueParts[0])] = strings.TrimSpace(itemValueParts[1])
+			} else {
+				errors = append(errors, fmt.Errorf("unrecognized item in prerelease: %v", itemValue))
+			}
+		}
+		if itemMap["version"] == "" {
+			errors = append(errors, fmt.Errorf("prerelease item %s is missing version info", prereleaseName))
+		}
+		if itemMap["featureGate"] == "" {
+			errors = append(errors, fmt.Errorf("prerelease item %s is missing feature gate info", prereleaseName))
+		}
+
+		fv := fieldVersion{
+			prerelease:  prereleaseName,
+			version:     itemMap["version"],
+			featureGate: itemMap["featureGate"],
+		}
+		fieldVersions = append(fieldVersions, fv)
+	}
+
+	return fieldVersions, errors
+}
+
+// TODO: validate version value
+func (fv fieldVersion) validate() []error {
+	errors := []error{}
+	if !allowedPrereleaseNames.Has(fv.prerelease) {
+		errors = append(errors, fmt.Errorf("unrecognized prerelease: %q. supported names are: %q", fv.prerelease, allowedPrereleaseNames.List()))
+	}
+	if fv.version == "" {
+		errors = append(errors, fmt.Errorf("prerelease item %s is missing version info", fv.prerelease))
+	}
+	if fv.featureGate == "" {
+		errors = append(errors, fmt.Errorf("prerelease item %s is missing feature gate info", fv.prerelease))
+	}
+	return errors
+}
+
+func validateFieldVersions(fieldVersions []fieldVersion) []error {
+	errors := []error{}
+	for _, fv := range fieldVersions {
+		if err := fv.validate(); err != nil {
+			errors = append(errors, err...)
+		}
+	}
+	return errors
+}
+
+// emit prints the fieldVersion, can be called on a nil fieldVersion (emits nothing)
+func (fv *fieldVersion) emit(g openAPITypeWriter) {
+	if fv == nil {
+		return
+	}
+	g.Do("map[string]interface{}{\n", nil)
+	if fv.prerelease != "" {
+		g.Do("\"prerelease\": \"$.$\",\n", fv.prerelease)
```

Comment:
will these become `x-kubernetes-...` extensions? They probably should.

Reply:
They will be part of `x-kubernetes-field-version` (this is a temporary name for the POC): https://github.com/kubernetes/kubernetes/pull/99307/commits/dd74c23bd3f2366f41defb3b37e62fdb12a33e5a